src/librbd/io/AioCompletion.cc

```diff
@@ -34,10 +34,8 @@ namespace io {
 int AioCompletion::wait_for_complete() {
   tracepoint(librbd, aio_wait_for_complete_enter, this);
   {
-    std::unique_lock<std::mutex> locker(lock);
-    while (state != AIO_STATE_COMPLETE) {
-      cond.wait(locker);
-    }
+    waiting_for_complete.store(true, std::memory_order_release);
+    state.wait(AIO_STATE_COMPLETE, std::memory_order_acquire);
```

Comment:
> Looks like there is a unit test failure sorry, debugging...

The signature is
```
void wait( T old, std::memory_order order =
                      std::memory_order::seq_cst ) const noexcept;
```
and it blocks if `old` matches the current value.  So it looks like you end up blocking when `state` is already `AIO_STATE_COMPLETE` instead of the other way around.

Reply:
Yes I've just realised that and I'm testing the fix now.

Reply:
So that was a silly mistake on my part, sorry - I misunderstood the atomic wait function as "wait until the state is equal to this value", rather than "wait until the state is NOT equal to this value" (I am a C++ newbie if you can't tell!).

I've just changed it to AIO_STATE_PENDING and ran all the RBD unit tests locally with no errors, so that seems to have fixed it. Is there an edge case though where the state is AIO_STATE_CALLBACK, or does that not matter since I only call notify_all after it gets set to AIO_STATE_COMPLETE?

Reply:
> Is there an edge case though where the state is AIO_STATE_CALLBACK, or does that not matter since I only call notify_all after it gets set to AIO_STATE_COMPLETE?

This is a good question!  I was going to answer "yes", because a user could call `AioCompletion::wait_for_complete()` while the state is `AIO_STATE_CALLBACK` (i.e. while callbacks are being run) and we don't want to return prematurely in that case.  There was an explicit fix to that effect, see commit 77f7c1cf829e5417eb356ebf3bf1fbc8ff7a25a5.

However, looking at today's code, the distinction between `AIO_STATE_CALLBACK` and `AIO_STATE_COMPLETE` doesn't seem to be material anymore:
- The entire point of  commit 77f7c1cf829e5417eb356ebf3bf1fbc8ff7a25a5 was this bit in `AioCompletion::start_op()`, but it was rewritten in commit 002afa0fe37544ab4eb15f973e00068e1240c0ef:

  ```
      Mutex::Locker locker(lock);
      if (state == STATE_PENDING && !async_op.started()) {
        async_op.start_op(*ictx);
      }
  ```

- The same commit 002afa0fe37544ab4eb15f973e00068e1240c0ef changed `AioCompletion::is_complete()` to test for `!= AIO_STATE_PENDING` instead of `== AIO_STATE_COMPLETE`.  No way to tell whether it was intended or a typo, but I think that it's a mistake that needs to be fixed because it made `AioCompletion::wait_for_complete()` and `AioCompletion::is_complete()` inconsistent with each other -- their meaning of "complete" is different in today's code.

With the above in mind, it seems like we can go ahead and revert state to a bool, fixing that mistake and simplifying at the same time:

- `done` would be set to true where state is set to `AIO_STATE_COMPLETE` today
-  you would be able to call `done.wait(false, ...)` without worrying about edge cases here
- `AioCompletion::is_complete()` would return `done`

What do you think?

Reply:
Yes that sounds good to me, I just searched the code for AIO_STATE_CALLBACK and can't find anywhere it is actually used anymore so was about to suggest the same.